File: python/semx/gxe.py
from typing import Any, Dict, Optional, Tuple, Union

import numpy as np
import pandas as pd

from .model import Model


def gxe_model(
    formula: str,
    data: pd.DataFrame,
    genotype: str,
    environment: str,
    genomic: Optional[Union[np.ndarray, Dict[str, Any]]] = None,
    family: str = "gaussian",
) -> Tuple[Model, pd.DataFrame]:
    """
    Builds a Genotype-by-Environment (GxE) model from a high-level specification.

    This builder simplifies the construction of Mixed Models for Multi-Environment
    Trials (MET). It assumes the data is in "long" format (one observation per
    genotype per environment).

    The constructed model includes:
      1. Fixed effects specified in ``formula``.
      2. A random intercept for ``genotype`` (G).
      3. A random intercept for ``environment`` (E).
      4. A random intercept for the GxE interaction (G:E).

    If ``genomic`` is provided, the G term uses the provided relationship matrix.
    The GxE term currently defaults to a diagonal (independent) structure.

    Args:
        formula: A string specifying the fixed effects (e.g., "yield ~ 1").
        data: A pandas DataFrame containing the data.
        genotype: The name of the column containing genotype identifiers.
        environment: The name of the column containing environment identifiers.
        genomic: Optional. A numeric matrix (numpy array) representing the
            genomic relationship matrix (GRM) for the genotypes.
        family: The outcome family (default: "gaussian").

    Returns:
        A tuple (model, data) where:
        - model: A :class:`semx.Model` instance ready for fitting.
        - data: The input DataFrame, potentially augmented with an interaction column.
    """
    # Parse response variable
    if "~" not in formula:
        raise ValueError("Formula must contain '~'")
    response = formula.split("~")[0].strip()

    # Create interaction column if needed
    interaction_col = f"{genotype}:{environment}"
    # We modify a copy to avoid side effects on the user's dataframe
    data = data.copy()
    if interaction_col not in data.columns:
        data[interaction_col] = data[genotype].astype(str) + ":" + data[environment].astype(str)

    # Encode grouping variables to integers (0-based indices)
    # This is required because the C++ core expects numeric data.
    # We use sort=True to ensure deterministic mapping (e.g. matching sorted GRM).
    if data[genotype].dtype == object or data[genotype].dtype.name == "category":
        codes, uniques = pd.factorize(data[genotype], sort=True)
        data[genotype] = codes
        if genomic is not None and genomic.shape[0] != len(uniques):
            raise ValueError(
                f"Genomic matrix dimension ({genomic.shape[0]}) does not match number of genotypes ({len(uniques)})"
            )

    if data[environment].dtype == object or data[environment].dtype.name == "category":
        data[environment] = pd.factorize(data[environment], sort=True)[0]

    if data[interaction_col].dtype == object or data[interaction_col].dtype.name == "category":
        data[interaction_col] = pd.factorize(data[interaction_col], sort=True)[0]

    # Inject intercept column explicitly since we are bypassing the formula parser's auto-injection
    if "_intercept" not in data.columns:
        data["_intercept"] = 1.0

    # Construct random effects list
    random_effects = []
    covariances = []
    genomic_args = {}

    # 1. Genotype Random Effect
    re_g = {
        "name": f"u_{genotype}",
        "variables": [genotype, "_intercept"],  # [grouping, design_var]
        "group": genotype,
        # Covariance will be set below
    }
    
    if genomic is not None:
        # Define genomic covariance
        cov_name = "K_g"
        # semx.Model expects genomic dict: name -> {markers: ...}
        genomic_args[cov_name] = {"markers": genomic}
        re_g["covariance"] = cov_name
        
        # We must also register the covariance definition
        covariances.append({
            "name": cov_name,
            "structure": "genomic",
            "dimension": 1
        })
    else:
        # Default diagonal covariance
        cov_name = f"cov_{genotype}"
        re_g["covariance"] = cov_name
        covariances.append({
            "name": cov_name,
            "structure": "diagonal",
            "dimension": 1
        })

    random_effects.append(re_g)

    # 2. Environment Random Effect
    cov_name_e = f"cov_{environment}"
    re_e = {
        "name": f"u_{environment}",
        "variables": [environment, "_intercept"],
        "group": environment,
        "covariance": cov_name_e
    }
    covariances.append({
        "name": cov_name_e,
        "structure": "diagonal",
        "dimension": 1
    })
    random_effects.append(re_e)

    # 3. Interaction Random Effect
    cov_name_gxe = f"cov_{interaction_col}"
    re_gxe = {
        "name": f"u_{interaction_col}",
        "variables": [interaction_col, "_intercept"],
        "group": interaction_col,
        "covariance": cov_name_gxe
    }
    covariances.append({
        "name": cov_name_gxe,
        "structure": "diagonal",
        "dimension": 1
    })
    random_effects.append(re_gxe)

    # Prepare families
    families = {response: family}
    # Register _intercept family to avoid ModelSpecificationError
    families["_intercept"] = "gaussian"

    # Construct Model
    # We pass the fixed effects formula.
    # We pass the explicit random effects.
    model = Model(
        equations=[formula],
        families=families,
        covariances=covariances,
        random_effects=random_effects,
        genomic=genomic_args if genomic_args else None
    )

    return model, data

File: Rpkg/semx/R/gxe.R
#' Genotype-by-Environment (GxE) Model
#'
#' Fits a Genotype-by-Environment (GxE) model using a mixed model framework.
#'
#' This function simplifies the construction of GxE models, supporting both
#' standard diagonal (independent) covariance structures and Kronecker product
#' structures ($K_G \otimes I_E$) for the interaction term.
#'
#' @param formula A formula specifying the fixed effects (e.g., `yield ~ 1`).
#' @param data A data frame containing the variables.
#' @param genotype Character string specifying the genotype column name.
#' @param environment Character string specifying the environment column name.
#' @param genomic Optional numeric matrix representing the genomic relationship matrix (GRM) or markers.
#'   If provided, the genotype random effect will use this covariance.
#' @param gxe_structure Character string: "diagonal" (default) or "kronecker".
#'   "kronecker" requires `genomic` to be provided.
#' @param family Outcome distribution family (default: "gaussian").
#' @param ... Additional arguments passed to `semx_fit`.
#'
#' @return A `semx_fit` object.
#' @export
semx_gxe <- function(formula, data, genotype, environment, genomic = NULL, gxe_structure = "diagonal", family = "gaussian", ...) {
  
  # Validate inputs
  if (!genotype %in% names(data)) stop(sprintf("Column '%s' not found in data", genotype))
  if (!environment %in% names(data)) stop(sprintf("Column '%s' not found in data", environment))
  
  # Parse formula to get response
  tf <- terms(as.formula(formula))
  vars <- attr(tf, "variables")
  resp_idx <- attr(tf, "response")
  if (resp_idx == 0) stop("Formula must have a response variable.")
  response <- as.character(vars[[resp_idx + 1]])
  
  # Prepare data copy
  df <- data
  
  # Factorize grouping variables to 0-based integers for C++
  # We use factor() to ensure consistent ordering
  
  # Genotype
  if (!is.numeric(df[[genotype]])) {
    f_g <- factor(df[[genotype]])
    # Sort levels if they are not ordered? 
    # Python implementation used sort=True. R factor levels are sorted by default if character.
    # But if they are already factor, we keep levels.
    # To match Python behavior (and potentially sorted GRM), we might want to force sort.
    # But user might have provided GRM matching the factor levels.
    # Let's assume user knows what they are doing with factor levels vs GRM rows.
    df[[genotype]] <- as.integer(f_g) - 1L
    n_g <- nlevels(f_g)
  } else {
    # Already numeric? Assume 0-based or 1-based?
    # C++ expects 0-based.
    # If user passes integers, we assume they are indices.
    # But safer to re-factorize to be sure.
    f_g <- factor(df[[genotype]])
    df[[genotype]] <- as.integer(f_g) - 1L
    n_g <- nlevels(f_g)
  }
  
  # Environment
  if (!is.numeric(df[[environment]])) {
    f_e <- factor(df[[environment]])
    df[[environment]] <- as.integer(f_e) - 1L
    n_e <- nlevels(f_e)
  } else {
    f_e <- factor(df[[environment]])
    df[[environment]] <- as.integer(f_e) - 1L
    n_e <- nlevels(f_e)
  }
  
  # Interaction
  interaction_col <- paste0(genotype, "_x_", environment)
  
  # Lists for semx_model
  covariances <- list()
  random_effects <- list()
  genomic_args <- list()
  
  # 1. Genotype Random Effect
  re_g_name <- paste0("u_", genotype)
  
  if (!is.null(genomic)) {
    cov_name_g <- "K_g"
    genomic_args[[cov_name_g]] <- list(markers = genomic)
    # Note: semx_model handles adding this to covariances if in genomic args?
    # Actually semx_model implementation adds genomic entries to covariances list.
    # So we don't need to add it manually to covariances list.
    
    re_g <- list(
      name = re_g_name,
      variables = c(genotype, "_intercept"),
      covariance = cov_name_g
    )
  } else {
    cov_name_g <- paste0("cov_", genotype)
    covariances[[length(covariances) + 1]] <- list(
      name = cov_name_g,
      structure = "diagonal",
      dimension = 1L
    )
    re_g <- list(
      name = re_g_name,
      variables = c(genotype, "_intercept"),
      covariance = cov_name_g
    )
  }
  random_effects[[length(random_effects) + 1]] <- re_g
  
  # 2. Environment Random Effect
  cov_name_e <- paste0("cov_", environment)
  covariances[[length(covariances) + 1]] <- list(
    name = cov_name_e,
    structure = "diagonal",
    dimension = 1L
  )
  re_e <- list(
    name = paste0("u_", environment),
    variables = c(environment, "_intercept"),
    covariance = cov_name_e
  )
  random_effects[[length(random_effects) + 1]] <- re_e
  
  # 3. Interaction Random Effect
  if (gxe_structure == "kronecker") {
    if (is.null(genomic)) stop("Kronecker GxE structure requires 'genomic' matrix.")
    
    # Create interaction index: g * n_e + e
    # Assuming K_g is outer, I_e is inner.
    # df[[genotype]] and df[[environment]] are 0-based.
    df[[interaction_col]] <- df[[genotype]] * n_e + df[[environment]]
    
    cov_name_ge <- "K_ge"
    cov_name_ie <- "I_e"
    
    # Define I_e
    covariances[[length(covariances) + 1]] <- list(
      name = cov_name_ie,
      structure = "identity",
      dimension = n_e
    )
    
    # Define K_ge
    covariances[[length(covariances) + 1]] <- list(
      name = cov_name_ge,
      structure = "kronecker",
      dimension = n_g * n_e,
      components = list(
          list(id = "K_g"),
          list(id = cov_name_ie)
      )
    )
    
    re_ge <- list(
      name = paste0("u_", interaction_col),
      variables = c(interaction_col, "_intercept"),
      covariance = cov_name_ge
    )
    random_effects[[length(random_effects) + 1]] <- re_ge
    
  } else {
    # Diagonal
    # Create interaction factor
    int_fac <- factor(paste(df[[genotype]], df[[environment]], sep = ":"))
    df[[interaction_col]] <- as.integer(int_fac) - 1L
    
    cov_name_gxe <- paste0("cov_", interaction_col)
    covariances[[length(covariances) + 1]] <- list(
      name = cov_name_gxe,
      structure = "diagonal",
      dimension = 1L
    )
    
    re_gxe <- list(
      name = paste0("u_", interaction_col),
      variables = c(interaction_col, "_intercept"),
      covariance = cov_name_gxe
    )
    random_effects[[length(random_effects) + 1]] <- re_gxe
  }
  
  # Prepare families
  families <- setNames(family, response)
  # Add intercept family if needed? semx_model handles it if we pass families.
  # But we need to ensure _intercept is registered if used.
  # semx_model does this if it sees _intercept in variables.
  
  # Construct model
  model <- semx_model(
    equations = c(formula),
    families = families,
    covariances = covariances,
    random_effects = random_effects,
    genomic = genomic_args,
    data = df
  )
  
  # Fit model
  fit <- semx_fit(model, df, ...)
  
  return(fit)
}
